### PR TITLE
Adding log to print all configuration flags at start of gcs fuse mount

### PR DESCRIPTION
**Pull request by ankitaluthra1**
### Description
Uses flagStorage struct to get all flags which are being used in gcsfuse and add a info log while starting gcsfuse

### Link to the issue in case of a bug fix.
[issue-link](https://b.corp.google.com/issues/315990701)

### Testing details
1. Manual - Done on Local
2. Unit tests - Added unit test for logic of aggregated log message containing all flags
3. Integration tests - NA

### CODE REVIEW

**Comment by raj-prince** [comment]
1. In this PR, we are only logging passed flags, right? We also need to print the config flags.
2. Could you please update the PR description with the newly added logs?

**Comment by ankitaluthra1** [comment]
Sample log (For now I have kept "" in the log as I am not sure removing " from logs will result in any formatting errors for any customers, let me know if we need to remove "" from log statement)
`{"time":{"timestampSeconds":1703683455,"timestampNanos":1703683455611934894},"severity":"INFO","msg":"Loading config for mount flags {\"AppName\":\"\",\"Foreground\":true,\"ConfigFile\":\"\",\"MountOptions\":{},\"DirMode\":493,\"FileMode\":420,\"Uid\":-1,\"Gid\":-1,\"ImplicitDirs\":false,\"OnlyDir\":\"\",\"RenameDirLimit\":0,\"CustomEndpoint\":null,\"BillingProject\":\"\",\"KeyFile\":\"\",\"TokenUrl\":\"\",\"ReuseTokenFromUrl\":true,\"EgressBandwidthLimitBytesPerSecond\":-1,\"OpRateLimitHz\":-1,\"SequentialReadSizeMb\":200,\"MaxRetrySleep\":30000000000,\"StatCacheCapacity\":4096,\"StatCacheTTL\":60000000000,\"TypeCacheTTL\":60000000000,\"HttpClientTimeout\":0,\"MaxRetryDuration\":-1000000000,\"RetryMultiplier\":2,\"LocalFileCache\":false,\"TempDir\":\"\",\"ClientProtocol\":\"http1\",\"MaxConnsPerHost\":100,\"MaxIdleConnsPerHost\":100,\"EnableNonexistentTypeCache\":false,\"StackdriverExportInterval\":0,\"OtelCollectorAddress\":\"\",\"LogFile\":\"\",\"LogFormat\":\"json\",\"ExperimentalEnableJsonRead\":false,\"DebugFuseErrors\":true,\"DebugFuse\":false,\"DebugFS\":false,\"DebugGCS\":false,\"DebugHTTP\":false,\"DebugInvariants\":false,\"DebugMutex\":false}"}
{"time":{"timestampSeconds":1703683455,"timestampNanos":1703683455611975904},"severity":"INFO","msg":"Loading config from config-file {\"CreateEmptyFile\":false,\"Severity\":\"INFO\",\"Format\":\"json\",\"FilePath\":\"\",\"LogRotateConfig\":{\"MaxFileSizeMB\":512,\"BackupFileCount\":10,\"Compress\":true}}"}`

**Comment by ankitaluthra1** [comment]
> 1. In this PR, we are only logging passed flags, right? We also need to print the config flags.
> 2. Could you please update the PR description with the newly added logs?
> 3. Lint is failing.

Added config file flag as well
Added sample logs in comment
Lint is fixed

File: main.go [hunk 1]
```diff
@@ -224,6 +234,8 @@ func runCLIApp(c *cli.Context) (err error) {
 	}
 
 	logger.Infof("Start gcsfuse/%s for app %q using mount point: %s\n", getVersion(), flags.AppName, mountPoint)
+	logger.Infof("Loading config for mount flags %s", stringify(*flags))
```
Comment (raj-prince) [inline main.go:237]: nit: instead we can write better like, 1GCSFuse mount command flags: %s`

File: main_test.go [hunk 2]
```diff
@@ -72,3 +75,64 @@ func (t *MainTest) TestGetUserAgentWhenMetadataImageTypeEnvVarSetAndAppNameNotSe
...
+				BackupFileCount: 2,
+				Compress:        true,
+			},
+		},
+	}
+
+	actual := stringify(mountConfig)
+
+	expected := "{\"CreateEmptyFile\":false,\"Severity\":\"TRACE\",\"Format\":\"\",\"FilePath\":\"\\\"path\\\"to\\\"file\\\"\",\"LogRotateConfig\":{\"MaxFileSizeMB\":2,\"BackupFileCount\":2,\"Compress\":true}}"
+	AssertEq(strings.TrimSpace(expected), strings.TrimSpace(actual))
```
Comment (raj-prince) [inline main_test.go:98]: Just curious, why do we need TrimSpace here?

File: main.go [hunk 3]
```diff
@@ -185,6 +186,15 @@ func populateArgs(c *cli.Context) (
 	return
 }
 
+func stringify(input any) string {
+	inputBytes, err := json.Marshal(input)
+
+	if err != nil {
+		logger.Errorf("Error in stringify %v", err)
```
Comment (raj-prince) [inline main.go:193]: We should log it as Warning, if it's not a fatal error.

File: main.go [hunk 4]
```diff
@@ -185,6 +186,15 @@ func populateArgs(c *cli.Context) (
 	return
 }
 
+func stringify(input any) string {
```
Comment (raj-prince) [inline main.go:189]: Can we move this method in util?
internal/util/util.go

File: main_test.go [hunk 5]
```diff
@@ -72,3 +75,64 @@ func (t *MainTest) TestGetUserAgentWhenMetadataImageTypeEnvVarSetAndAppNameNotSe
...
+		Value: "example",
+	}
+
+	actual := stringify(customInstance)
+
+	expected := ""
+	AssertEq(strings.TrimSpace(expected), strings.TrimSpace(actual))
+}
+
+type CustomType struct {
```
Comment (raj-prince) [inline main_test.go:131]: nit: Is it mandatory to make this object public? Can't we keep `customType`?

File: main.go (hunk 1 above)
Comment (ankitaluthra1) [reply main.go]: Updated logs with above text

File: main_test.go [hunk 6]
```diff
@@ -72,3 +75,64 @@ func (t *MainTest) TestGetUserAgentWhenMetadataImageTypeEnvVarSetAndAppNameNotSe
...
+
+	expected := "{\"AppName\":\"\",\"Foreground\":false,\"ConfigFile\":\"\",\"MountOptions\":{\"1\":\"one\",\"2\":\"two\",\"3\":\"three\"},\"DirMode\":0,\"FileMode\":0,\"Uid\":0,\"Gid\":0,\"ImplicitDirs\":false,\"OnlyDir\":\"\",\"RenameDirLimit\":0,\"CustomEndpoint\":null,\"BillingProject\":\"\",\"KeyFile\":\"\",\"TokenUrl\":\"\",\"ReuseTokenFromUrl\":false,\"EgressBandwidthLimitBytesPerSecond\":0,\"OpRateLimitHz\":0,\"SequentialReadSizeMb\":10,\"MaxRetrySleep\":0,\"StatCacheCapacity\":0,\"StatCacheTTL\":0,\"TypeCacheTTL\":0,\"HttpClientTimeout\":0,\"MaxRetryDuration\":0,\"RetryMultiplier\":0,\"LocalFileCache\":false,\"TempDir\":\"\",\"ClientProtocol\":\"http4\",\"MaxConnsPerHost\":0,\"MaxIdleConnsPerHost\":0,\"EnableNonexistentTypeCache\":false,\"StackdriverExportInterval\":0,\"OtelCollectorAddress\":\"\",\"LogFile\":\"\",\"LogFormat\":\"\",\"ExperimentalEnableJsonRead\":false,\"DebugFuseErrors\":false,\"DebugFuse\":false,\"DebugFS\":false,\"DebugGCS\":false,\"DebugHTTP\":false,\"DebugInvariants\":false,\"DebugMutex\":false}"
+	AssertEq(strings.TrimSpace(expected), strings.TrimSpace(actual))
+}
+
+func (t *MainTest) TestStringifyShouldReturnEmptyStringWhenMarshalErrorsOut() {
+
```
Comment (raj-prince) [inline main_test.go:120]: nit: extra new line.

**Review by raj-prince — CHANGES_REQUESTED** [review]
Few minor comments.

File: main_test.go (hunk 2 above)
Comment (ankitaluthra1) [reply main_test.go]: its not needed now actually, have removed it (added it earlier was testing random data without quotes)

File: main.go (hunk 3 above)
Comment (ankitaluthra1) [reply main.go]: done.. changed it to warn

File: main.go (hunk 4 above)
Comment (ankitaluthra1) [reply main.go:189]: it was only used in one file hence kept it here for now, will it be okay if we move it to utility when more than one file uses it

File: main_test.go (hunk 5 above)
Comment (ankitaluthra1) [reply main_test.go:131]: made it private

File: main.go (hunk 4 above)
Comment (ankitaluthra1) [reply main.go:189]: additionally, I won't be able to test flagStorage example as flagStorage is private to main package

File: main_test.go (hunk 6 above)
Comment (ankitaluthra1) [reply main_test.go]: removed

File: main.go (hunk 4 above)
Comment (ankitaluthra1) [reply main.go]: have moved Stringify to util package

File: internal/util/util_test.go [hunk 7]
```diff
@@ -157,3 +159,53 @@ func (t *UtilTest) TestResolveConfigFilePaths() {
...
 	ExpectEq(filepath.Join(homeDir, "test.txt"), mountConfig.LogConfig.FilePath)
 }
+
+func (t *UtilTest) TestStringifyShouldReturnAllFieldsPassedInCustomObjectAsMarshalledString() {
+	sampleMap := map[string]int{
+		"1": 1,
+		"2": 2,
+		"3": 3,
+	}
+
```
Comment (raj-prince) [inline internal/util/util_test.go:169]: nit: Please remove new line, in unit test we use newline to separate arrange, act and assert statment.

File: internal/util/util.go [hunk 8]
```diff
@@ -76,3 +77,13 @@ func ResolveConfigFilePaths(config *config.MountConfig) (err error) {
 	}
 	return
 }
+
+func Stringify(input any) string {
```
Comment (raj-prince) [inline internal/util/util.go:81]: nit: We can write a doc here, which helps in understanding the method just hovering over the usage.

File: internal/util/util_test.go [hunk 9]
```diff
@@ -157,3 +159,53 @@ func (t *UtilTest) TestResolveConfigFilePaths() {
...
+	actual := Stringify(customInstance)
+
+	expected := ""
+	AssertEq(strings.TrimSpace(expected), strings.TrimSpace(actual))
+}
+
+type customTypeForSuccess struct {
+	Value       string
```
Comment (raj-prince) [inline internal/util/util_test.go:197]: Just curious, what happens if we make the attribute private?

File: internal/util/util_test.go [hunk 10]
```diff
@@ -157,3 +159,53 @@ func (t *UtilTest) TestResolveConfigFilePaths() {
...
+
+func (t *UtilTest) TestStringifyShouldReturnEmptyStringWhenMarshalErrorsOut() {
+	customInstance := customTypeForError{
+		value: "example",
+	}
+
+	actual := Stringify(customInstance)
+
+	expected := ""
+	AssertEq(strings.TrimSpace(expected), strings.TrimSpace(actual))
```
Comment (raj-prince) [inline internal/util/util_test.go:193]: Why TrimSpace is required here?

**Review by raj-prince — CHANGES_REQUESTED** [review]
LGTM, Added few nits.

File: internal/util/util_test.go (hunk 9 above)
Comment (ankitaluthra1) [reply internal/util/util_test.go:197]: it was unable to marshal.. I verified

File: internal/util/util.go (hunk 8 above)
Comment (ankitaluthra1) [reply internal/util/util.go:82]: done

File: internal/util/util_test.go (hunk 10 above)
Comment (ankitaluthra1) [reply internal/util/util_test.go:192]: its typo... not required, have removed it

File: internal/util/util_test.go (hunk 7 above)
Comment (ankitaluthra1) [reply internal/util/util_test.go]: done

File: internal/util/util.go [hunk 11]
```diff
@@ -76,3 +77,14 @@ func ResolveConfigFilePaths(config *config.MountConfig) (err error) {
 	}
 	return
 }
+
+// Returns empty String in case of failure in marshalling the passed input.
```
Comment (raj-prince) [inline internal/util/util.go:81]: nit: Stringify marshals an object (only exported attribute) to a JSON string. If marshalling fails, it returns an empty string.

**Review by raj-prince — APPROVED** [review]
LGTM, minor doc change.

File: internal/util/util.go (hunk 11 above)
Comment (ankitaluthra1) [reply internal/util/util.go:81]: done

**Review by raj-prince — APPROVED** [review]

**Pull request merged**